docs/operation/upgrade.rst

```diff
@@ -0,0 +1,110 @@
+Upgrade Guide
+=============
+This document describes a complete and reliable upgrade path for **MetalK8s**
+and all the components that make up the stack.
+
+.. important::
+
+    - Upgrading MetalK8s is a manual process and as a result,
+      the following outlined steps must be met to the later in other to achieve
+      a successful upgrade.
+    - The output of the below-provided commands solely depends on the particular
+      MetalK8s setup you have in place.
+
+.. warning::
+    For any cluster upgrade from MetalK8s v2.0 to v2.1 and on-ward, **Kubelet**
+    must be upgraded before making any upgrades to the salt-master. Once this
+    is done, an upgrade is guaranteed to proceed in a seamless manner.
+
+Supported Versions
+******************
+
+This guide currently supports MetalK8s upgrades for the following versions.
+Please refer to the release notes for more information.
+
+    - Upgrade 2.0.0 to 2.1.0
+    - Upgrade 2.0.0 to 2.2.0
+    - Upgrade 2.1.0 to 2.2.0
+
+
+Upgrade Steps
+*************
+This section describes the logical steps for Upgrading MetalK8s.
+
+#. Connect to the :term:`Bootstrap node`.
+
+#. Download the new `.ISO` from `Sources <http://scality.com>`_.
+
+#. Edit the `bootstrap.yaml` file located in `/etc/metalk8s/bootstrap.yaml`
+   then add the new MetalK8s ISO mount point.
+
+   .. code::
+
+      products:
+        metalk8s:
+        - /srv/scality/metalk8s-2.0.0-dev
+        - /srv/scality/metalk8s-<new_version>     <=== New line to add
+
+
+#. Mount the new MetalK8s ISO.
+
+    ::
+
+        [root@bootstrap vagrant]# mkdir -p /srv/scality/metalk8s-<new_version>/
+
+    ::
+
+        [root@bootstrap vagrant]# mount metalk8s.iso /srv/scality/metalk8s-<new_version>/
+
+#. Verify that the current MetalK8s version is mounted.
+
+    ::
+
+        salt-call state.sls metalk8s.products.mounted saltenv=metalk8s-<current_version>
+
+            Summary for local
+            ------------
+            Succeeded: 5 (changed=1)
+            Failed:    0
+            ------------
+            Total states run:     5
+
+#. Advertise the new version and make it available to the entire cluster.
+
+    ::
+
+        salt-call --local state.sls metalk8s.products saltenv=metalk8s-<new_version> pillar="{'metalk8s': {'endpoints': $(salt-call --out txt pillar.get metalk8s:endpoints | cut -c 8-)}}"
+
+#. From the :term:`Bootstrap node`, connect to the :term:`Salt Master`.
+
+    ::
+
+        kubectl -n kube-system exec -it salt-master-bootstrap bash
```

Comment:
I know it's the simplest way but if we connect to the master using `kubectl` as we upgrade kubelet we will get disconnected from the salt-master so I suggest to use `crictl` (btw we will have the same problem if we upgrade containerd...).

I think in the futur use `salt-api` will be the best choice. (Maybe through the UI)

Reply:
Yeah, if this is possible via the salt-api like you have suggested. But you leave me with no choice here.

What to add or remove since kubectl  or crictl will behave the same way??

Reply:
I think crictl is still better for the moment as between metalk8s version, for the moment, we not really change containerd version (btw we take the last version available in centos repos at build time, so may change between two build) but kubelet we always change it between 2 minor of metalk8s

Reply:
Also here you assume the hostname of the bootstrap node is `bootstrap`.
You should use filtering with "labels" like `crictl exec -it $(crictl ps -q --label io.kubernetes.container.name=salt-master) bash`